timeutil/schedule.go

```diff
@@ -54,34 +91,168 @@ func ParseTime(s string) (t TimeOfDay, err error) {
 	if err != nil {
 		return t, fmt.Errorf("cannot parse %q: %s", m[2], err)
 	}
+	if t.Hour == 24 {
+		if t.Minute > 0 {
+			return t, fmt.Errorf("cannot parse %q: not a valid time", s)
+		}
+	}
 	return t, nil
 }
 
-// Schedule defines a start and end time in which events should run.
+// MonthWeekday is a helper type to represent a Weekday that occurs on a
+// specific week in a month, eg. wed4 -> Wednesday in the 4th week of the month
+type MonthWeekday int
+
+const (
+	// sane default
+	monthWeekdayNone MonthWeekday = 0
+	monthWeekdayMin  MonthWeekday = 1
+	// not possible to have given day appear more than 5 times a month
+	monthWeekdayMax MonthWeekday = 5
+	// indicates the last weekday of a month, eg. the last Monday
+	monthWeekdayLast = monthWeekdayMax
+)
+
+func (m MonthWeekday) String() string {
+	if m == monthWeekdayNone {
+		return ""
+	}
+	return strconv.Itoa(int(m))
+}
+
+// Schedule defines a start and end time and an optional weekday in which
+// events should run.
 type Schedule struct {
 	Start TimeOfDay
 	End   TimeOfDay
 
+	// weekday the schedule starts on
 	Weekday string
+	// weekday the schedule ends on, can be empty if the same as Weekday
+	WeekdayEnd string
+
+	// specific weekday of the month, eg. the first Monday, the 2nd Tuesday, the
+	// last Monday
+	MonthWeekday MonthWeekday
+	// indicates when the weekday-of-the-month schedule ends, can be set to
+	// monthWeekdayNone if the same as MonthWeekday
+	MonthWeekdayEnd MonthWeekday
+
+	// if true then even time will be randomized
+	Randomize bool
 }
 
 func (sched *Schedule) String() string {
-	return fmt.Sprintf("%s-%s", sched.Start, sched.End)
+	sep := "-"
+	if sched.Randomize {
+		sep = "~"
+	}
+
+	buf := &bytes.Buffer{}
+
+	if sched.Weekday != "" {
+		fmt.Fprintf(buf, "%s%s", sched.Weekday, sched.MonthWeekday)
+	}
+	if sched.WeekdayEnd != "" {
+		mwe := sched.MonthWeekdayEnd
+		if mwe == monthWeekdayNone && sched.MonthWeekday != monthWeekdayNone {
+			mwe = sched.MonthWeekdayEnd
+		}
+		fmt.Fprintf(buf, "-%s%s", sched.WeekdayEnd, mwe)
+	}
+
+	if sched.Weekday != "" {
+		fmt.Fprint(buf, ",")
+	}
+	fmt.Fprintf(buf, "%s", sched.Start)
+	if sched.End != sched.Start {
+		fmt.Fprint(buf, sep)
+		fmt.Fprintf(buf, "%s", sched.End)
+	}
+	return buf.String()
+}
+
+// isLastWeekdayInMonth returns true it a.Weekday() is the last weekday
+// occurring this a.Month(), eg. check is Feb 25 2017 is the last Saturday of
+// February.
+func isLastWeekdayInMonth(a time.Time) bool {
+	month := a.Month()
+	isLast := false
+	t := a
+
+	// try a week from now
+	t = t.Add(7 * 24 * time.Hour)
+
+	if t.Month() != month {
+		// if hit a different month then current weekday is the last in
+		// this month
+		isLast = true
+	}
```

Comment:
More compact: `return t.Month() != month`, and you can get rid of isLast variable.

Reply:
Must have been a mental lapse on my side.